File: reader_writer(mutex).py
import threading
import time
import random

# Shared resource (initialized to None to indicate no data available)
shared_resource = None

# Mutex locks
mutex = threading.Lock()          # Protects the reader count
resource_access = threading.Lock()  # Protects the shared resource
writer_done = threading.Event()  # Event to signal when the writer has written

# Reader count
read_count = 0

# Reader function
def reader(reader_id):
    global read_count
    while True:
        # Wait until a writer has written data
        writer_done.wait()

        # Entry section
        with mutex:
            read_count += 1
            if read_count == 1:  # First reader locks the resource
                resource_access.acquire()

        # Critical section (Reading)
        print(f"Reader {reader_id} is reading the value: {shared_resource}")
        time.sleep(random.uniform(0.1, 1))  # Simulate reading time

        # Exit section
        with mutex:
            read_count -= 1
            if read_count == 0:  # Last reader unlocks the resource
                resource_access.release()

        # Simulate delay before reading again
        time.sleep(random.uniform(0.5, 2))

# Writer function
def writer(writer_id):
    global shared_resource
    while True:
        # Entry section
        resource_access.acquire()

        # Critical section (Writing)
        shared_resource = random.randint(1, 100)  # Generate a new value
        print(f"Writer {writer_id} updated the value to: {shared_resource}")
        writer_done.set()  # Signal that the writer has written
        time.sleep(random.uniform(0.5, 1))  # Simulate writing time

        # Exit section
        resource_access.release()

        # Simulate delay before writing again
        time.sleep(random.uniform(1, 3))

# Main function
if _name_ == "_main_":
    # Create reader and writer threads
    readers = [threading.Thread(target=reader, args=(i,)) for i in range(3)]  # 3 readers
    writers = [threading.Thread(target=writer, args=(i,)) for i in range(1)]  # 1 writer

    # Start threads
    for r in readers:
        r.start()
    for w in writers:
        w.start()

    # Stop the program after 10 seconds
    time.sleep(10)
    print("Program terminated.")
    for r in readers:
        r.join(timeout=1)
    for w in writers:
        w.join(timeout=1)

File: reader_writer(mutex).c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

// Mutexes
pthread_mutex_t resource_mutex = PTHREAD_MUTEX_INITIALIZER; // Protects the shared resource
pthread_mutex_t read_count_mutex = PTHREAD_MUTEX_INITIALIZER; // Protects read_count
pthread_mutex_t op_count_mutex = PTHREAD_MUTEX_INITIALIZER;  // Protects operation_count

int read_count = 0;  // Tracks the number of readers accessing the resource
int shared_data = 0; // Shared resource
int operation_count = 20; // Total operations to perform (read + write)

// Writer function
void* writer(void* arg) {
    int writer_id = *(int*)arg; // Correctly dereference the argument

    while (1) {
        pthread_mutex_lock(&op_count_mutex);
        if (operation_count <= 0) {
            pthread_mutex_unlock(&op_count_mutex);
            break;
        }
        operation_count--;
        pthread_mutex_unlock(&op_count_mutex);

        pthread_mutex_lock(&resource_mutex); // Exclusive access for writers

        // Writing to the shared resource
        shared_data += 1;
        printf("Writer %d wrote: %d\n", writer_id, shared_data);

        pthread_mutex_unlock(&resource_mutex); // Release the resource
        sleep(rand() % 2 + 1); // Simulate writing time
    }

    return NULL;
}

// Reader function
void* reader(void* arg) {
    int reader_id = *(int*)arg; // Correctly dereference the argument

    while (1) {
        pthread_mutex_lock(&op_count_mutex);
        if (operation_count <= 0) {
            pthread_mutex_unlock(&op_count_mutex);
            break;
        }
        operation_count--;
        pthread_mutex_unlock(&op_count_mutex);

        // Entry Section
        pthread_mutex_lock(&read_count_mutex);
        read_count++;
        if (read_count == 1) {
            pthread_mutex_lock(&resource_mutex); // First reader locks the resource
        }
        pthread_mutex_unlock(&read_count_mutex);

        // Reading the shared resource
        printf("Reader %d read: %d\n", reader_id, shared_data);

        // Exit Section
        pthread_mutex_lock(&read_count_mutex);
        read_count--;
        if (read_count == 0) {
            pthread_mutex_unlock(&resource_mutex); // Last reader releases the resource
        }
        pthread_mutex_unlock(&read_count_mutex);

        sleep(rand() % 2 + 1); // Simulate reading time
    }

    return NULL;
}

// Main function
int main() {
    pthread_t readers[3], writers[2]; // Arrays for reader and writer threads

    // Dynamically allocate IDs to avoid race conditions
    int* reader_ids[3];
    int* writer_ids[2];

    for (int i = 0; i < 3; i++) {
        reader_ids[i] = malloc(sizeof(int));
        *reader_ids[i] = i + 1;
    }
    for (int i = 0; i < 2; i++) {
        writer_ids[i] = malloc(sizeof(int));
        *writer_ids[i] = i + 1;
    }

    // Create writer threads
    for (int i = 0; i < 2; i++) {
        pthread_create(&writers[i], NULL, writer, writer_ids[i]);
    }

    // Create reader threads
    for (int i = 0; i < 3; i++) {
        pthread_create(&readers[i], NULL, reader, reader_ids[i]);
    }

    // Join threads
    for (int i = 0; i < 2; i++) {
        pthread_join(writers[i], NULL);
        free(writer_ids[i]); // Free allocated memory
    }
    for (int i = 0; i < 3; i++) {
        pthread_join(readers[i], NULL);
        free(reader_ids[i]); // Free allocated memory
    }

    return 0;
}
